pkg/controllers/clusterresourceplacement/controller.go

```diff
@@ -192,25 +201,69 @@ func (r *Reconciler) deleteRedundantSchedulingPolicySnapshots(ctx context.Contex
 	if err != nil {
 		return err
 	}
-
-	crpKObj := klog.KObj(crp)
 	// respect the revisionHistoryLimit field
+	revisionLimit := parseRevisionLimit(crp)
+	if len(sortedList.Items) < revisionLimit {
+		return nil
+	}
+	for i := 0; i <= len(sortedList.Items)-revisionLimit; i++ { // need to reserve one slot for the new snapshot
+		if err := r.Client.Delete(ctx, &sortedList.Items[i]); err != nil && !errors.IsNotFound(err) {
+			klog.ErrorS(err, "Failed to delete clusterSchedulingPolicySnapshot", "clusterResourcePlacement", klog.KObj(crp), "clusterSchedulingPolicySnapshot", klog.KObj(&sortedList.Items[i]))
+			return controller.NewAPIServerError(false, err)
+		}
+	}
+	return nil
+}
+
+func parseRevisionLimit(crp *fleetv1beta1.ClusterResourcePlacement) int {
 	revisionLimit := fleetv1beta1.RevisionHistoryLimitDefaultValue
 	if crp.Spec.RevisionHistoryLimit != nil {
 		revisionLimit = *crp.Spec.RevisionHistoryLimit
 		if revisionLimit <= 0 {
-			err := fmt.Errorf("invalid clusterResourcePlacement %s: invalid revisionHistoryLimit %d", crpKObj, revisionLimit)
-			klog.ErrorS(controller.NewExpectedBehaviorError(err), "Invalid revisionHistoryLimit value and using default value instead", "clusterResourcePlacement", crpKObj)
+			err := fmt.Errorf("invalid clusterResourcePlacement %s: invalid revisionHistoryLimit %d", crp.Name, revisionLimit)
+			klog.ErrorS(controller.NewUnexpectedBehaviorError(err), "Invalid revisionHistoryLimit value and using default value instead", "clusterResourcePlacement", klog.KObj(crp))
 			// use the default value instead
 			revisionLimit = fleetv1beta1.RevisionHistoryLimitDefaultValue
 		}
 	}
-	if len(sortedList.Items) < int(revisionLimit) {
+	return int(revisionLimit)
+}
+
+// deleteRedundantResourceSnapshots handles multiple snapshots in a group.
+func (r *Reconciler) deleteRedundantResourceSnapshots(ctx context.Context, crp *fleetv1beta1.ClusterResourcePlacement) error {
+	sortedList, err := r.listSortedResourceSnapshots(ctx, crp)
+	if err != nil {
+		return err
+	}
+
+	// respect the revisionHistoryLimit field
+	revisionLimit := parseRevisionLimit(crp)
```

Comment:
This limit is parsed multiple times in the code; since both snapshot types use the same limit, it might be better to extract it out to a common place.